chain/chain/src/store.rs

```diff
@@ -887,81 +887,72 @@ impl ChainStoreAccess for ChainStore {
-
-        // It is a lifetime workaround. We cannot return `&mut changes_per_key_prefix.filter_map(...`
-        // as that is a temporary object created there with `filter_map`, and you cannot leak a
-        // reference to a temporary object.
-        let mut changes_per_exact_key;
-
-        let changes_per_key: &mut dyn Iterator<Item = _> = match state_changes_request {
-            // These request types are expected to match the key exactly:
-            StateChangesRequest::AccountChanges { .. }
-            | StateChangesRequest::SingleAccessKeyChanges { .. }
-            | StateChangesRequest::CodeChanges { .. } => {
-                changes_per_exact_key = changes_per_key_prefix.filter_map(|change| {
-                    let (key, state_changes) = match change {
-                        Ok(change) => change,
-                        error => {
-                            return Some(error);
-                        }
-                    };
-                    if key.len() != data_key.len() {
-                        None
-                    } else {
-                        debug_assert_eq!(key.as_ref(), data_key.as_ref() as &[u8]);
-                        Some(Ok((key, state_changes)))
-                    }
-                });
-                &mut changes_per_exact_key
-            }
-
-            StateChangesRequest::AllAccessKeyChanges { .. }
-            | StateChangesRequest::DataChanges { .. } => &mut changes_per_key_prefix,
-        };
-
         Ok(match state_changes_request {
-            StateChangesRequest::AccountChanges { account_id, .. } => {
-                StateChanges::from_account_changes(changes_per_key, account_id)?
+            StateChangesRequest::AccountChanges { account_ids } => {
+                let mut changes = StateChanges::new();
+                for account_id in account_ids {
+                    let data_key = utils::KeyForAccount::new(account_id);
+                    let storage_key = KeyForStateChanges::new(&block_hash, data_key.as_ref());
+                    let changes_per_key = storage_key.find_exact_iter(&self.store);
+                    changes
+                        .extend(StateChanges::from_account_changes(changes_per_key, account_id)?);
+                }
+                changes
             }
-            StateChangesRequest::SingleAccessKeyChanges { account_id, .. }
-            | StateChangesRequest::AllAccessKeyChanges { account_id, .. } => {
-                let access_key_pk = match state_changes_request {
-                    StateChangesRequest::SingleAccessKeyChanges { access_key_pk, .. } => {
-                        Some(access_key_pk)
-                    }
-                    _ => None,
-                };
-                StateChanges::from_access_key_changes(changes_per_key, account_id, access_key_pk)?
+            StateChangesRequest::SingleAccessKeyChanges { account_ids, access_key_pk } => {
```

Comment:
This is strange API to have from normal perspective. Why would someone would need to track multiple accounts with only one access key. It means those accounts share the same access key. Which is terrible idea from pseudo-anonymity and security perspective.

I'd suggest to change it to provide pairs of `(account_id, access_key_pk)` for the API to make sense.

Reply:
Very good point! I will update the API.